File: app-presentation/src/main/java/com/zxjdev/smile/presentation/common/di/component/ApplicationComponent.java
package com.zxjdev.smile.presentation.common.di.component;

import android.app.Application;

import com.zxjdev.smile.presentation.common.SmileApplication;
import com.zxjdev.smile.presentation.common.di.module.ActivityBindingModule;
import com.zxjdev.smile.presentation.common.di.module.ApplicationModule;
import com.zxjdev.smile.presentation.common.di.module.AuthorizationDomainModule;
import com.zxjdev.smile.presentation.common.di.module.MomentDomainModule;
import com.zxjdev.smile.presentation.common.di.module.UserDomainModule;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;
import dagger.android.AndroidInjector;
import dagger.android.support.AndroidSupportInjectionModule;

@Singleton
@Component(modules = {
  AndroidSupportInjectionModule.class,
  ApplicationModule.class,
  AuthorizationDomainModule.class,
  MomentDomainModule.class,
  UserDomainModule.class,
  ActivityBindingModule.class
})
public interface ApplicationComponent extends AndroidInjector<SmileApplication> {

  @Component.Builder
  interface Builder {

    @BindsInstance
    ApplicationComponent.Builder application(Application application);

    ApplicationComponent build();
  }
}

File: app-presentation/src/main/java/com/zxjdev/smile/presentation/common/di/component/ApplicationComponent.kt
package com.zxjdev.smile.presentation.common.di.component

import android.app.Application
import com.zxjdev.smile.presentation.common.SmileApplication
import com.zxjdev.smile.presentation.common.di.module.*
import dagger.BindsInstance
import dagger.Component
import dagger.android.AndroidInjector
import dagger.android.support.AndroidSupportInjectionModule
import javax.inject.Singleton

@Singleton
@Component(modules = arrayOf(AndroidSupportInjectionModule::class,
        ApplicationModule::class,
        AuthorizationDomainModule::class,
        MomentDomainModule::class,
        UserDomainModule::class,
        ActivityBindingModule::class,
        PresentationModule::class))
interface ApplicationComponent : AndroidInjector<SmileApplication> {

    @Component.Builder
    interface Builder {

        @BindsInstance
        fun application(application: Application): ApplicationComponent.Builder

        fun build(): ApplicationComponent
    }
}
